rasterio/_warp.pyx

```diff
@@ -140,15 +130,30 @@ cdef GDALWarpOptions * create_warp_options(
 
     psWOptions.eResampleAlg = <GDALResampleAlg>resampling
 
+    # Assign nodata values.
+    # We don't currently support an imaginary component.
+
     psWOptions.padfSrcNoDataReal = <double*>CPLMalloc(src_count * sizeof(double))
     psWOptions.padfSrcNoDataImag = <double*>CPLMalloc(src_count * sizeof(double))
+
+    for i in range(src_count):
+        if src_nodata is not None:
+            psWOptions.padfSrcNoDataReal[i] = float(src_nodata)
+            psWOptions.padfSrcNoDataImag[i] = 0.0
+        else:
+            psWOptions.padfSrcNoDataReal[i] = -123456.789  # as in gdalwarp.
```

Comment:
The GDAL docs say the nodata arrays can be NULL, but it doesn't seem to be the case (TODO: ask on gdal-dev). In the gdalwarp code, a special nodata value that we hope never occurs is used.